libsolidity/analysis/NameAndTypeResolver.cpp

```diff
@@ -476,7 +476,13 @@ bool DeclarationRegistrationHelper::registerDeclaration(
 	Declaration const* shadowedDeclaration = nullptr;
 	if (_warnOnShadow && !name.empty() && _container.enclosingContainer())
 		for (auto const* decl: _container.enclosingContainer()->resolveName(name, true, true))
-			shadowedDeclaration = decl;
+			// Do not warn about functions shadowing a contract.
+			if (
+				!(dynamic_cast<ContractDefinition const*>(decl) &&
+				dynamic_cast<FunctionDefinition const*>(&_declaration) &&
+				name == decl->name())
+			)
+				shadowedDeclaration = decl;
```

Comment:
So only something that is a function can be shadowed by something that's not a contract with the same name?
What about ``contract C { uint msg; }``? Currently this warns: ``This declaration shadows a builtin symbol.``. And I think this comes from below in this function (I may be wrong) - does that still work? Is there a test for it?

Reply:
We have a test for the case you've mentioned: ``452_shadowing_builtins_with_storage_variables.sol``

Reply:
After we talked about that: I've improved the formatting to make the condition easier to understand 😄